launch_frontend/launch_frontend/parser.py

```diff
@@ -0,0 +1,150 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for Parser methods."""
+
+from typing import Optional
+from typing import Text
+
+import launch
+
+from pkg_resources import iter_entry_points
+
+from .entity import Entity
+from .expose import action_parse_methods, expose_action
+from .substitutions import default_substitution_interpolation
+
+interpolation_fuctions = {
+    entry_point.name: entry_point.load()
+    for entry_point in iter_entry_points('launch_frontend.interpolate_substitution')
+}
+
+extensions_loaded = False
+
+
+def load_parser_extensions():
+    for entry_point in iter_entry_points('launch_frontend.parser_extension'):
+        entry_point.load()
+
+
+def parse_action(entity: Entity) -> launch.Action:
+    """Parse an action, using its registered parsing method."""
+    if not extensions_loaded:
+        load_parser_extensions()
+    if entity.type_name not in action_parse_methods:
+        raise RuntimeError('Unrecognized entity of the type: {}'.format(entity.type_name))
+    return action_parse_methods[entity.type_name](entity)
+
+
+def parse_substitution(value: Text, frontend: Text) -> launch.SomeSubstitutionsType:
+    """Parse a substitution, using its registered parsing method."""
+    if not extensions_loaded:
+        load_parser_extensions()
+    if frontend in interpolation_fuctions:
+        return interpolation_fuctions[frontend](value)
+    else:
+        return default_substitution_interpolation(value)
+
+
+def parse_description(entity: Entity) -> launch.LaunchDescription:
+    """Parse a launch description."""
+    if entity.type_name != 'launch':
+        raise RuntimeError('Expected \'launch\' as root tag')
+    actions = [parse_action(child) for child in entity.children]
+    return launch.LaunchDescription(actions)
+
+
+def str_to_bool(string):
+    """Convert xs::boolean to python bool."""
+    if not string or string.lower() in ('0', 'false'):
+        return False
+    if string.lower() in ('1', 'true'):
+        return True
+    raise RuntimeError('Expected "true" or "false", got {}'.format(string))
+
+
+def load_optional_attribute(
+    kwargs: dict,
+    entity: Entity,
+    name: Text,
+    *,
+    subst: bool = True,
+    constructor_name: Optional[Text] = None
+):
+    """Load an optional attribute of `entity` named `name` in `kwargs`."""
+    attr = getattr(entity, name, None)
+    key = name
+    if constructor_name is not None:
+        key = constructor_name
+    if attr is not None:
+        if subst:
+            kwargs[key] = parse_substitution(attr, entity.frontend)
+        else:
+            kwargs[key] = attr
+
+
+@expose_action('executable')
+def parse_executable(entity: Entity):
+    """Parse executable tag."""
+    cmd = parse_substitution(entity.cmd, entity.frontend)
```

Comment:
@ivanpauno this is why the parsing functions would take a `Parser` instance in the design document.

Reply:
Thinking again about this, having different `Parser` instance is better.

- The base class should provide all the action and substitution parsing methods.
- The base class should provide a `parse_action` method.
- The base class should provide a `parse_substitution` default method. This method could be overridden in the derived class.
- The base class should provide a `parse_description` method.
- The base class should provide a `load` method, which infers the used frontend from the file extension (and maybe does a brute-force load) and call the `load` method of the derived class.
- The derived class should provide a `load` method for creating an `Entity` from a file.

What do you think?